File: C++/2-Medium/00155._最小栈/00155._最小栈.cpp
#include <iostream>
#include <stack>
using namespace std;

class MinStack {
private:
    stack<int> s; // 主栈，存储所有元素
    stack<int> minS; // 辅助栈，存储当前最小值

public:
    /** initialize your data structure here. */
    MinStack() {
        // 构造函数不需要初始化代码
    }

    void push(int x) {
        s.push(x); // 元素x入主栈
        // 如果辅助栈为空或x小于等于辅助栈顶元素，则x也入辅助栈
        if (minS.empty() || x <= minS.top()) {
            minS.push(x);
        }
    }

    void pop() {
        // 如果主栈顶元素等于辅助栈顶元素，辅助栈也要弹出
        if (s.top() == minS.top()) {
            minS.pop();
        }
        s.pop(); // 主栈弹出
    }

    int top() {
        return s.top(); // 返回主栈顶元素
    }

    int getMin() {
        return minS.top(); // 返回辅助栈顶元素，即当前最小值
    }
};

int main() {
    MinStack minStack;
    minStack.push(-2);
    minStack.push(0);
    minStack.push(-3);
    cout << "Min: " << minStack.getMin() << endl; // Returns -3
    minStack.pop();
    cout << "Top: " << minStack.top() << endl;    // Returns 0
    cout << "Min: " << minStack.getMin() << endl; // Returns -2
    return 0;
}

File: Python/2-Medium/00155._最小栈/00155._最小栈.py
from typing import List

class MinStack:
    def __init__(self):
        """
        初始化数据结构。主栈用于存储所有元素，辅助栈用于存储每个状态下的最小值。
        """
        self.stack = []  # 主栈
        self.min_stack = []  # 辅助栈

    def push(self, x: int) -> None:
        """
        如果辅助栈为空或新元素小于等于辅助栈顶元素，则同时将新元素压入辅助栈。
        """
        self.stack.append(x)
        if not self.min_stack or x <= self.min_stack[-1]:
            self.min_stack.append(x)

    def pop(self) -> None:
        """
        弹出主栈顶部元素，如果弹出的元素等于辅助栈顶部元素，则也弹出辅助栈顶部元素。
        """
        x = self.stack.pop()
        if x == self.min_stack[-1]:
            self.min_stack.pop()

    def top(self) -> int:
        """
        获取主栈顶部元素。
        """
        return self.stack[-1]

    def getMin(self) -> int:
        """
        获取辅助栈顶部元素，即当前最小值。
        """
        return self.min_stack[-1]

# 测试代码
min_stack = MinStack()
min_stack.push(-2)
min_stack.push(0)
min_stack.push(-3)
print("Min:", min_stack.getMin())  # 应返回 -3
min_stack.pop()
print("Top:", min_stack.top())     # 应返回 0
print("Min:", min_stack.getMin())  # 应返回 -2